pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```diff
@@ -0,0 +1,512 @@
+		// If the lease is not held by a CLE aware holder and the lease is not expired
+		// then PreferredHolder is not recognized and election is unnecessary since
+		// the holder will not voluntarily give up the lease.
+		// This may occur on version skew cases when CLE is first enabled.
+		if lease.Annotations[ElectedByAnnotationName] != controllerName && !isLeaseExpired(lease) {
+			return
+		}
+	}
+
+	c.electionQueue.Add(leaderLeaseID{namespace: lease.Namespace, name: lease.Name})
+}
+
+func (c *Controller) electionNeeded(candidates []*v1alpha1.LeaseCandidate, leaderLeaseID leaderLeaseID) (bool, error) {
+	lease, err := c.leaseInformer.Lister().Leases(leaderLeaseID.namespace).Get(leaderLeaseID.name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return false, fmt.Errorf("error reading lease")
+	} else if apierrors.IsNotFound(err) {
+		return true, nil
+	}
+	if isLeaseExpired(lease) || lease.Spec.HolderIdentity == nil || *lease.Spec.HolderIdentity == "" {
+		return true, nil
+	}
+
+	if len(candidates) == 0 {
+		return true, nil
+	}
+
+	prelimStrategy := pickBestStrategy(candidates)
+	if prelimStrategy != v1.OldestEmulationVersion {
+		return false, fmt.Errorf("strategy %s is not recognized by CLE", prelimStrategy)
+	}
+
+	prelimElectee := pickBestLeaderOldestEmulationVersion(candidates)
+	if lease != nil && lease.Spec.HolderIdentity != nil && prelimElectee.Name == *lease.Spec.HolderIdentity {
+		klog.V(2).Infof("Leader %s is already most optimal for lease %s %s", prelimElectee.Name, leaderLeaseID.namespace, leaderLeaseID.name)
+		return false, nil
+	}
+	return true, nil
+}
+
+// reconcileElectionStep steps through a step in an election.
+// A step looks at the current state of Lease and LeaseCandidates and takes one of the following action
+// - do nothing (because leader is already optimal or still waiting for an event)
+// - request ack from candidates (update LeaseCandidate PingTime)
+// - finds the most optimal candidate and elect (update the Lease object)
+// Instead of keeping a map and lock on election, the state is
+// calculated every time by looking at the lease, and set of available candidates.
+// PingTime + electionDuration > time.Now: We just asked all candidates to ack and are still waiting for response
+// PingTime + electionDuration < time.Now: Candidate has not responded within the appropriate PingTime. Continue the election.
+// RenewTime + 5 seconds > time.Now: All candidates acked in the last 5 seconds, continue the election.
+func (c *Controller) reconcileElectionStep(ctx context.Context, leaderLeaseID leaderLeaseID) (bool, error) {
+	now := time.Now()
+	klog.V(4).Infof("reconcileElectionStep %q %q", leaderLeaseID.namespace, leaderLeaseID.name)
+
+	candidates, err := c.listAdmissableCandidates(leaderLeaseID)
+	if err != nil {
+		return true, err
+	}
+
+	// Check if an election is really needed by looking at the current lease
+	// and set of candidates
+	reelection, err := c.electionNeeded(candidates, leaderLeaseID)
```

Comment:
reelection sounds like the same candidate is elected again. This is exactly NOT what you want, is it?

Reply:
renamed to needElection.